src/Uno.Core/Collections/ImmutableList.cs

```diff
@@ -24,109 +27,367 @@ namespace Uno.Collections
 	/// <summary>
 	/// An immutable list implementation, designed for safe concurrent access.
 	/// </summary>
-	public class ImmutableList<T>
+	/// <remarks>
+	/// It is possible to mutate the content of this immutable list. Not all
+	/// protections are in place to ensure a perfect immutability. Don't use it
+	/// if you are exposing it outside your code. For true immutability
+	/// protection, you should prefer those from Micrososft's Immutable Collections.
+	/// </remarks>
+	public class ImmutableList<T> : IImmutableList<T>
 	{
-		private T[] data;
-		private static ImmutableList<T> _empty = new ImmutableList<T>();
+		private readonly T[] _data;
+
+		/// <summary>
+		/// Provides an empty list
+		/// </summary>
+		public static ImmutableList<T> Empty { get; } = new ImmutableList<T>(new T[0], false);
 
 		/// <summary>
 		/// Creates an empty list
 		/// </summary>
 		public ImmutableList()
 		{
-			data = new T[0];
+			_data = Empty.Data;
 		}
 
 		/// <summary>
-		/// Provides an empty list
+		/// Initializes the list with the provided array.
 		/// </summary>
-		public static ImmutableList<T> Empty {
-			get { return _empty; }
+		/// <param name="data">An array as source</param>
+		/// <param name="copyData">If the array should be copied</param>
+		public ImmutableList(T[] data, bool copyData = false)
```

Comment:
Do we want to let this public ? If so, does the `copyData` should not be set to `true` by default ?

Reply:
copyData is false by default because of existing code using it right now - as it's the previous behavior.

@jeromelaban what do you think?